roofit/roofitcore/test/test_lib.h

```diff
@@ -0,0 +1,185 @@
+    {
+      std::ostringstream os;
+      os << "Uniform::u" << ix << "(x" << ix << ")";
+      w.factory(os.str().c_str());
+    }
+
+    // gather the observables in a list for data generation below
+    {
+      std::ostringstream os;
+      os << "x" << ix;
+      obs_set.add(*w.arg(os.str().c_str()));
+    }
+  }
+
+  RooArgSet pdf_set = w.allPdfs();
+
+  // create event counts for all pdfs
+  RooArgSet count_set;
+
+  // ... for the gaussians
+  for (unsigned ix = 0; ix < n; ++ix) {
+    std::stringstream os, os2;
+    os << "Nsig" << ix;
+    os2 << "#signal events comp " << ix;
+    RooRealVar a(os.str().c_str(), os2.str().c_str(), N_events/10, 0., 10*N_events);
+    w.import(a);
+    // gather in count_set
+    count_set.add(*w.arg(os.str().c_str()));
+  }
+  // ... and for the uniform background components
+  for (unsigned ix = 0; ix < n; ++ix) {
+    std::stringstream os, os2;
+    os << "Nbkg" << ix;
+    os2 << "#background events comp " << ix;
+    RooRealVar a(os.str().c_str(), os2.str().c_str(), N_events/10, 0., 10*N_events);
+    w.import(a);
+    // gather in count_set
+    count_set.add(*w.arg(os.str().c_str()));
+  }
+
+  RooAddPdf* sum = new RooAddPdf("sum", "gaussians+uniforms", pdf_set, count_set);
+  w.import(*sum);  // keep sum around after returning
+
+  // --- Generate a toyMC sample from composite PDF ---
+  RooDataSet *data = sum->generate(obs_set, N_events);
+
+  std::unique_ptr<RooAbsReal> nll {sum->createNLL(*data)};
+
+  // set values randomly so that they actually need to do some fitting
+  for (unsigned ix = 0; ix < n; ++ix) {
+    {
+      std::ostringstream os;
+      os << "m" << ix;
+      dynamic_cast<RooRealVar *>(w.arg(os.str().c_str()))->setVal(RooRandom::randomGenerator()->Gaus(0, 2));
+    }
+    {
+      std::ostringstream os;
+      os << "s" << ix;
+      dynamic_cast<RooRealVar *>(w.arg(os.str().c_str()))->setVal(0.1 + abs(RooRandom::randomGenerator()->Gaus(0, 2)));
+    }
+  }
+
+  // gather all values of parameters, pdfs and nll here for easy
+  // saving and restoring
+  std::unique_ptr<RooArgSet> all_values = std::make_unique<RooArgSet>(pdf_set, count_set, "all_values");
+  all_values->add(*nll);
+  all_values->add(*sum);
+  for (unsigned ix = 0; ix < n; ++ix) {
+    {
+      std::ostringstream os;
+      os << "m" << ix;
+      all_values->add(*w.arg(os.str().c_str()));
+    }
+    {
+      std::ostringstream os;
+      os << "s" << ix;
+      all_values->add(*w.arg(os.str().c_str()));
+    }
+  }
+
+  return std::make_tuple(std::move(nll), sum, data, std::move(all_values));
+}
+
+
+class Hex {
+public:
+  explicit Hex(double n) : number_(n) {}
+  operator double() const { return number_; }
+  bool operator==(const Hex& other) {
+    return double(*this) == double(other);
+  }
+
+private:
+  double number_;
+};
+
+::std::ostream& operator<<(::std::ostream& os, const Hex& hex) {
+  return os << std::hexfloat << double(hex) << std::defaultfloat;  // whatever needed to print bar to os
+}
```

Comment:
Is this a relic from older times?

Reply:
Yep, removed it! I just use `printf("%a", some_float)` now.

Reply:
Ah, actually, this was a helper class for use in googletest checks, you could just surround the numbers in equality checking asserts with `Hex()` and gtest would print them in hexadecimal instead of regular.